### PR TITLE
Albop/compile static

### PR DESCRIPTION
**Pull request by albop**
@sglyon, This PR proposes new methods to generate functions. It is intentionally meant to coexist with the current implementation and reuses some of the tricks. Here is the spirit:
- a `FlatFunctionFactory` replaces the `FunctionFactory` object. This one drops the distinction between grouped and non-grouped arguments, assignments and non-assignments. Also, it deals only with standard Symbol and expressions, assuming manipulations on timed expressions have already been done. One goal is this new object is precisely to make it easier to compile functions.
- derivation can be done (at first order) with respect to any combination of the arguments. `f(Val(0), x,y,z,p)` returns residuals,  `f(Val(1), x,y,z,p)` returns derivative with first arguments (`x`) while `f((Val(0),Val(1)), x,y,z,p)` returns both packed in a tuple.
- functions operate on points  (`SVector`) and lists of points ( `Vector{SVector}`) and return points, lists of points or lists of jacobian (`Vector{SMatrix}`). If output arguments are supplied, it should not allocate memory. It some prelinary, there seem to be big performance gains (x4 without multithreading).
- function are generated in three layers:
    - the kernel (`gen_kernel`) : takes `SVector`s  as input and returns a tuple of `SVector` or `SMatrix`. This function defines the core dimensions. It contains the same information as the computation *kernel* of python's guvectorize functions. It looks like a natural input to other modifications like autodiff or alike.
    - the gufun (`gen_gufun`): vectorizes the kernel. Currently does it only with one additional dim (i.e. matrices can be passed where vectors are expected), but it would be useful to at least extend to 2 dims.
    - generated (`gen_gengufun`): writes a generated gufun which takes an additional static type parameter to determine which derivatives are to be computed as in `f((Val(0), Val(2)), x,y,z,p)`. This type parameter is currently independent from the `Der{x}` type. The `@generated` keyword puts many (artificial?) constraints on what can be done (no eval, generated code must be pure...). I tried to limit the impact of these constraints on the other functions.

I have tried to keep things as simple as possible and left aside the generation of higher order functions. I am not sure whether it should be done in another set of functions or not. There is still work to be done here, like to compute the derivatives when there are definitions or recursive blocks.

Also this depends on SymEngine. The approach consisting in pre-generating the first order functions makes no sense when we have to consider all possible combinations. Funnily enough the tests pass without it, as apprently Calculus.jl doesn't always need to use eval. (here `genfun((Val(0),Val(1)),x,y,z,p))` works but not `genfun((Val(0),Val(2)),x,y,z,p))` ...)

### CODE REVIEW

**Comment by albop** [comment]
Two updates from this morning:
- dispatch type can be passed to generating function
- derivatives are computed using chain rules (and should be correct). The kernel code is something like:
```
function myfun(x::SVector{1, Float64}, y::SVector{3, Float64}, z::SVector{2, Float64}, p::SVector{1, Float64})
    _a_m1_ = x[1]
    _a__0_ = y[1]
    _b__0_ = y[2]
    _c__0_ = y[3]
    _c__1_ = z[1]
    _d__1_ = z[2]
    _u_ = p[1]
    _foo__0_ = log(_a__0_) + _b__0_ / (_a_m1_ / (1 - _c__0_))
    ∂_foo__0__∂_a_m1_ = 0.0
    ∂_foo__0__∂_c__0_ = -_b__0_ / _a_m1_
    ∂_foo__0__∂_a__0_ = _a__0_ ^ -1
    ∂_foo__0__∂_b__0_ = (1 - _c__0_) / _a_m1_
    _bar__0_ = _c__1_ + _u_ * _d__1_
    ∂_bar__0__∂_u_ = _d__1_
    ∂_bar__0__∂_a_m1_ = 0.0
    ∂_bar__0__∂_c__1_ = 1
    ∂_bar__0__∂_d__1_ = _u_
    oo_1 = SVector(_foo__0_, _bar__0_)
    oo_2 = SMatrix{2, 1}(∂_foo__0__∂_a_m1_, ∂_bar__0__∂_a_m1_)
    res_ = (oo_1, oo_2)
    return res_
end
```

**Comment by sglyon** [comment]
Ok, it seems like the code here has settled down. 

I also have more time now as some deadlines are behind me. I’m planning on doing a review on Monday.

**Comment by albop** [comment]
That's OK. There are three small changes upcoming:
- I might add a compatibility call to generated functions in the form `transition!` without output argument coming first, (rather than transition(..., out=nothing)). Not strictly necessary as I have a branch in Dolo without these calls.
- surprisingly the compatibility matrix calls (which copy and transpose before doing anything) are faster than the older functions. This is almost certainly due to how Julia deals with views. So the plan to use a temporary optional `orientation` keyword could also be used here (as a keyword to the generated function).
- there is a function solve_dependency which is slightly broken. I'm working on it in a different branch (`simplify_symbolic`). This is relevant to construct the FlatFunctionFactory eventually, but I don't think it is even used there.

**Comment by sglyon** [comment]
So the very latest commit allows us to call a generated gu-function using syntax `genfufun(Val{(0, 1, 3}), args...)` for levels and derivatives wrt arguments 1 and 3. I think this is a nice improvement over what we were able to do previosuly.

File: REQUIRE [hunk 1]
```diff
@@ -2,3 +2,4 @@ julia 0.5
 DataStructures
 Compat 0.18
 Calculus
+StaticArrays
```
[the remark this answers is not in the document]
Comment (albop) [reply REQUIRE]: Yes, that makes sense. I tried SymEngine on Windows and it seems to work.

File: src/Dolang.jl [hunk 2]
```diff
@@ -35,6 +36,8 @@ end::Bool
 if HAVE_SYMENGINE
     import SymEngine
     deriv(eq::SymEngine.Basic, x) = SymEngine.diff(eq, x)
+    # somebody will probably object to it:
+    deriv(eq::Expr, x::Symbol) = parse(string(SymEngine.diff(SymEngine.Basic(eq), SymEngine.Basic(x))))
```
[the remark this answers is not in the document]
Comment (albop) [reply src/Dolang.jl]: I didn't find how to pass an expression to symengine directly, which explains the conversion to/from strings.

File: src/compiler_new.jl [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+    return diff_eqs
+
+end
+
+"""
+Create a non allocating kernel from the function factory.
+
+`fff`: assumed to be a `FlatFunctionFactory` object with empty preamble.
+`diff`: index of variables to differentiate with or list of indices of variables positions.
```
[the remark this answers is not in the document]
Comment (albop) [reply src/compiler_new.jl]: It's just unclear, it should be `Union{Vector{<:Integer},Integer}`

**Comment by albop** [comment]
I like the last commit. Actually, I wondered whether it was possible to write smthg like `DDer{0,1,2}` like one does `Tuple{0,1,2}`. Haven't found. Also we could think of a way to merge the different syntaxes: `Der{k}` for `k-tth` order approx w.r.t. first argument while `DDer{x_1, ... x_k}`  would differentiate variables with indices x_1, ... x_k.  Mostly esthetical though so really not urgent.

**Comment by sglyon** [comment]
Once we update the REQUIRE file and remove any notion of Calculus.jl I think this is ready to be merged.

**Comment by albop** [comment]
OK. done.

**Pull request merged**